### PR TITLE
feat(kubernetes): Improve the security of the kubernetes/helm charts

### PR DESCRIPTION
**Pull request by shakti-garg-saxo**
## Checklist
- [x] The PR conforms to DataHub's [Contributing Guideline](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [x] Docs related to the changes have been added/updated (if applicable)

### CODE REVIEW

**Comment by shakti-garg-saxo** [comment]
This PR fulfils the feature request #1747

File: contrib/kubernetes/datahub/Chart.yaml [hunk 1]
```diff
@@ -4,24 +4,24 @@ description: A Helm chart for LinkedIn DataHub
 type: application
 # This is the chart version. This version number should be incremented each time you make changes
 # to the chart and its templates, including the app version.
-version: 0.0.1
+version: 0.1.0
```
Comment (mars-lan) [inline contrib/kubernetes/datahub/Chart.yaml:7]: I have limited knowledge in helm, but any reason why we're not aligning this version with individual dependency versions (i.e. `0.2.0`)?

**Comment by mars-lan** [comment]
@jsotelo can you also help review this?

File: contrib/kubernetes/datahub/charts/datahub-gms/templates/deployment.yaml [hunk 2]
```diff
@@ -37,70 +37,37 @@ spec:
...
-                secretKeyRef:
-                  name: {{ include "datahub-gms.fullname" . }}-secret
-                  key: ebean.datasource.username
+              value: "{{ .Values.global.sql.datasource.username }}"
             - name: EBEAN_DATASOURCE_PASSWORD
               valueFrom:
                 secretKeyRef:
-                  name: {{ include "datahub-gms.fullname" . }}-secret
-                  key: ebean.datasource.password
+                  name: "{{ .Values.global.sql.datasource.password.secretRef }}"
```
Comment (jsotelo) [inline contrib/kubernetes/datahub/charts/datahub-gms/templates/deployment.yaml:44]: Another approach that I've seen others embrace is this: https://github.com/helm/helm/issues/2780#issuecomment-403307198 . It also allows dynamic environment variables which is good because Spring Kafka allows for dynamic properties via `spring.kafka.properties.*` see https://docs.spring.io/spring-boot/docs/current/reference/html/appendix-application-properties.html#integration-properties

File: contrib/kubernetes/datahub/Chart.yaml (hunk 1 above)
Comment (shakti-garg-saxo) [reply contrib/kubernetes/datahub/Chart.yaml:7]: Each chart (4 charts for gms services and 1 composing chart called datahub) is an independent entity in itself and can maintain its own versions. The composing chart, "datahub" can subscribe to any version of sub charts and thus can maintain its own versioning.
For example, I did some minor change only to chart for "datahub-gms" and thus upgrade its version to 0.2.1. Other sub charts version will still be 0.2.0. The composing chart can be upgrade to 0.1.1.
Similarly, I can add another sub-chart for any new service in future. It's first version will be 0.1.0, though rest of the sub-charts will be much higher.
In short, it is like Maven dependency management, transitive dependency version has no correlation to composing dependency.

Thanks.

Comment (mars-lan) [reply contrib/kubernetes/datahub/Chart.yaml:7]: Got it. As long as this is following the helm chart convention which sounds like it is.

File: contrib/kubernetes/datahub/charts/datahub-mae-consumer/templates/deployment.yaml [hunk 3]
```diff
@@ -33,45 +33,24 @@ spec:
           imagePullPolicy: {{ .Values.image.pullPolicy }}
           env:
             - name: KAFKA_BOOTSTRAP_SERVER
-              valueFrom:
```
[the remark this answers is not in the document]
Comment (Nidhimnn) [reply contrib/kubernetes/datahub/charts/datahub-mae-consumer/templates/deployment.yaml:36]: I understand the underlying advantage of using configmap but it also adds on to security risk wherein we are not aware of changes made to configmap resources. Also, for higher envs like prod and other client-facing envs, I(CD) want to be aware of such changes and hence can version the deployment based on changes so that in future if these new changes in configmap is responsible for any failure, we can directly roll-back.

File: contrib/kubernetes/datahub/charts/datahub-gms/templates/deployment.yaml (hunk 2 above)
Comment (shakti-garg-saxo) [reply contrib/kubernetes/datahub/charts/datahub-gms/templates/deployment.yaml:44]: Thanks @jsotelo 
I agree with this approach for definingdynamic properties. The properties and secrets defined in values.yaml till now are mandatory arguments to the chart.
IMO, values.yaml is like an input API to the chart, mix of mandatory and optional parameters. mandatory args should be explicitly mentioned and optional ones can be defined using the dynamic properties(kerberos/sasl configs) like extraEnv, extraVolumes, extraSecrets (https://hackernoon.com/the-art-of-the-helm-chart-patterns-from-the-official-kubernetes-charts-8a7cafa86d12)
Needless to say, mentioning them explicitly also improves the documentation of the charts.

File: contrib/kubernetes/datahub/charts/datahub-mae-consumer/templates/deployment.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (jsotelo) [reply contrib/kubernetes/datahub/charts/datahub-mae-consumer/templates/deployment.yaml:36]: I'm not sure I understand how configmaps are less secure. If others don't have an issue with baking in environmental variables in the deployment, it's not a deal breaker for me.

File: contrib/kubernetes/datahub/charts/datahub-gms/templates/deployment.yaml (hunk 2 above)
Comment (jsotelo) [reply contrib/kubernetes/datahub/charts/datahub-gms/templates/deployment.yaml:44]: I agree, explicit environment variables are the way to go b/c they are self documenting. I was suggesting that in addition to them, we add a way to inject dynamic ones to allow user to configure `spring.kafka.properties` fully. Otherwise I am not sure how we would be able to configure things like `SPRING_KAFKA_PROPERTIES_SASL_KERBEROS_SERVICE_NAME` and `SPRING_KAFKA_PROPERTIES_SECURITY_PROTOCOL`, etc. Trying to keep in-sync with kafka source code and specify all possibilities is not practical.

Comment (shakti-garg-saxo) [reply contrib/kubernetes/datahub/charts/datahub-gms/templates/deployment.yaml:48]: Agree. Thanks for pointing this out and I have added a commit(e395b4b) to address dynamic properties by adding three variables, extraEnvs, extraVolumes and extraVolumeMounts to the values.yaml for each chart.
We are using the same for providing kerberos related properties like  SPRING_KAFKA_PROPERTIES_SECURITY_PROTOCOL, mounting keytab file and kerberos configuration file(krb5.conf) .
Sorry for my miss in cherrypicking it to PR branch earlier.

**Comment by shakti-garg-saxo** [comment]
Added two more commits:
1) add extraEnvs, extraVolumes and extraVolumeMounts to enable adding dynamic properties and mounting configs or secrets
2) As by default imageTag is "latest", default value of pullPolicy should be "Always" to override k8s node's local docker cache.

**Comment by mars-lan** [comment]
@jsotelo do you still have any concerns with this PR. If not I'm happy to merge it in.

**Comment by jsotelo** [comment]
> @jsotelo do you still have any concerns with this PR. If not I'm happy to merge it in.

From a code review perspective everything looks good. I wanted to try out an actual helm deployment into our dev environment but I'm not sure I'll have time to do that this week. :(

**Comment by mars-lan** [comment]
@shakti-garg-saxo if you've tested this end-to-end I'll merge it in for now. @jsotelo can follow up with a PR for any missed corner cases.

**Comment by shakti-garg-saxo** [comment]
@mars-lan Yes, we have tested this end-to-end and currently, running this in our dev environment. 

There might be some more security requirements coming up once we inch closer to prod. We will raise separate PRs for those if any. 

Sorry for replying late.

**Comment by SheetalPratik09** [comment]
Thanks Mars ans Shakti,

**Comment by shakti-garg** [comment]
Resolves #1747
